source/common/ssl/connection_impl.cc

```diff
@@ -134,47 +134,56 @@ Network::ConnectionImpl::IoResult ConnectionImpl::doWriteToSocket() {
+
+    uint64_t inner_bytes_written = 0;
+    for (uint64_t i = 0; i < num_slices; i++) {
+      // SSL_write() requires that if a previous call returns SSL_ERROR_WANT_WRITE, we need to call
+      // it again with the same parameters. Most implementations keep track of the last write size.
+      // In our case we don't need to do that because: a) SSL_write() will not write partial
+      // buffers. b) We only move() into the write buffer, which means that it's impossible for a
+      // particular chain to increase in size. So as long as we start writing where we left off we
+      // are guaranteed to call SSL_write() with the same parameters.
+      int rc = SSL_write(ssl_.get(), slices[i].mem_, slices[i].len_);
+      conn_log_trace("ssl write returns: {}", *this, rc);
+      if (rc > 0) {
+        inner_bytes_written += rc;
+        total_bytes_written += rc;
+      } else {
+        int err = SSL_get_error(ssl_.get(), rc);
+        switch (err) {
+        case SSL_ERROR_WANT_WRITE:
+          keep_writing = false;
+          break;
+        case SSL_ERROR_WANT_READ:
+        // Renegotiation has started. We don't handle renegotiation so just fall through.
+        default:
+          drainErrorQueue();
+          return {PostIoAction::Close, total_bytes_written};
+        }
 
-  uint64_t num_slices = write_buffer_.getRawSlices(nullptr, 0);
-  Buffer::RawSlice slices[num_slices];
-  write_buffer_.getRawSlices(slices, num_slices);
-
-  uint64_t bytes_written = 0;
-  for (uint64_t i = 0; i < num_slices; i++) {
-    // SSL_write() requires that if a previous call returns SSL_ERROR_WANT_WRITE, we need to call
-    // it again with the same parameters. Most implementations keep track of the last write size.
-    // In our case we don't need to do that because: a) SSL_write() will not write partial buffers.
-    // b) We only move() into the write buffer, which means that it's impossible for a particular
-    // chain to increase in size. So as long as we start writing where we left off we are guaranteed
-    // to call SSL_write() with the same parameters.
-    int rc = SSL_write(ssl_.get(), slices[i].mem_, slices[i].len_);
-    conn_log_trace("ssl write returns: {}", *this, rc);
-    if (rc > 0) {
-      bytes_written += rc;
-    } else {
-      int err = SSL_get_error(ssl_.get(), rc);
-      switch (err) {
-      case SSL_ERROR_WANT_WRITE:
         break;
-      case SSL_ERROR_WANT_READ:
-      // Renegotiation has started. We don't handle renegotiation so just fall through.
-      default:
-        drainErrorQueue();
-        return {PostIoAction::Close, bytes_written};
       }
-
-      break;
     }
-  }
 
-  if (bytes_written > 0) {
-    write_buffer_.drain(bytes_written);
+    if (inner_bytes_written > 0) {
+      write_buffer_.drain(inner_bytes_written);
+    }
   }
 
-  return {PostIoAction::KeepOpen, bytes_written};
+  return {PostIoAction::KeepOpen, total_bytes_written};
```

Comment:
Unfortunately there is no global way to do this, since the code currently assumes that it returns all slices, or the size of the slices array required to hold them all. I will put in a TODO to figure this out and clean this up globally. I want to get a fix for this crash out since we saw this in production.